### PR TITLE
CI: Node.js 20 GHA updates

### PR DESCRIPTION
**Pull request by hannahramadan**
Update the following actions to newer versions that use Node.js 20+

- actions/checkout - closes [#2598](https://github.com/newrelic/newrelic-ruby-agent/issues/2598)
- actions/upload-artifact - closes [#2599](https://github.com/newrelic/newrelic-ruby-agent/issues/2599)
- actions/download-artifact - closes [#2600](https://github.com/newrelic/newrelic-ruby-agent/issues/2600)
- actions/cache - closes [#2601](https://github.com/newrelic/newrelic-ruby-agent/issues/2601)
- actions/github-script - closes [#2602](https://github.com/newrelic/newrelic-ruby-agent/issues/2602)
- adriangl/check-new-commits-action - closes [#2604](https://github.com/newrelic/newrelic-ruby-agent/issues/2604)
- softprops/action-gh-release - closes [#2605](https://github.com/newrelic/newrelic-ruby-agent/issues/2605)
- .github/actions/issue_closer - closes [#2606](https://github.com/newrelic/newrelic-ruby-agent/issues/2606)

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (hannahramadan) [inline .github/workflows/ci.yml:108]: Upgrading from v3 to v4 requires artifacts to have different names as they are immutable, unlike in v3. 

Migration guide: https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md

**Review by kaylareopelle — DISMISSED** [review]

File: .github/versions.yml
Comment (fallwith) [reply .github/versions.yml:11]: I think we should stick with commits that at least belong to the repo and not to a fork. A feature branch might be ok, but I think we should draw the line at forks.

**Review by fallwith — CHANGES_REQUESTED** [review]
I think we should remove any references to forks and mark the relevant Node 20 upgrade effort(s) as blocked until the PRs are merged into the main (non-fork) repos.

File: .github/versions.yml
Comment (fallwith) [reply .github/versions.yml:11]: The obvious exception would be a `newrelic` owned fork...

File: .github/workflows/ci.yml
Comment (fallwith) [reply .github/workflows/ci.yml:108]: Will this change impact anything such as the code coverage stitcher or Stoat? If so, we should document those with new GitHub Issues and weigh whether or not we should upgrade this one or hold off until the child issues are addressed.

Comment (hannahramadan) [reply .github/workflows/ci.yml:108]: @fallwith I'm not seeing any issues with code coverage being stitched (example run w/code coverage at the bottom: https://github.com/newrelic/newrelic-ruby-agent/actions/runs/10167351518).

I'm not however sure how Stoat might be impacted (I'm not sure how stoat would use this code). Is this something you know off the top of your head?

File: .github/versions.yml
Comment (hannahramadan) [reply .github/versions.yml:11]: I see how a fork might be a little too saucy. I don't really know how to get newrelic to fork something lol! I'll revert this. Are we comfortable creating a separate backlog issue for this one and revisiting? We could even move this action into our `.github` folder like how I was thinking with [Create local technote-space/workflow-conclusion-action gha](https://github.com/newrelic/newrelic-ruby-agent/issues/2790)

File: .github/workflows/ci.yml
Comment (fallwith) [reply .github/workflows/ci.yml:108]: I was coming at this from the perspective of all artifact usage involving two steps: an upload step, and a download step; with the download step being optional in some cases where we might want to store results for a potential manual review that never ends up being needed.

The line I was commenting on is impacting the upload step by changing the name of the upload. So that made me wonder if the corresponding download step(s) would be impacted and need to be updated to be aware of the new name.

It looks like this repo's own downloading logic is tolerant to the name change because this `ci.yml` file's `simplecov` job seems to just grab ALL artifacts from the workflow without caring about the names. The actual stitcher at `lib/tasks/coverage_report.rake` seems to make use of splats (`*`) to not care about the name change either as long as the names still start with `coverage_`. So we're good there with this repo.

Stoat has fully three separate subcommands that all make use of the `fetch_artifact` helper, so I wondered if it might be affected as a downloader of artifacts uploaded by this repo's actions.

It seems like for this `ci.yml` file, it focuses only on `coverage-report-combined` which this PR does not rename, so we are ok.

File: .github/versions.yml
Comment (fallwith) [reply .github/versions.yml:11]: > Are we comfortable

I'm not sure of what the deadline is and what happens when after the deadline. If the deadline is far away, we're comfortable. If the deadline being reached simply causes the action to be forced to run on a Node v20 runtime, we're comfortable. If the deadline is soon and/or it involves the action failing instead of simply running under v20, then we'll have to evaluate urgency.

I see at least these 5 ways forward (not listed in any intentional order) - all of which I am happy to respect the group consensus on:

- Override me and go with this PR as-is. Perhaps throw me a bone and at least put a "TODO: " prefix in the .yml file's comment line and create a backlogged issue for the TODO.
- Create the separate backlog issue and revisit as you mentioned.
- Point to the fork instead of the original repo. It's really the mixing of 2 repos that I have an issue with. Unless we have a particularly strict process for allowlisting new actions repos, we could just point to the fork. Unfortunately the rake task isn't smart enough to handle switching repos and the workflow lines would need to be manually changed.
- Fork the action under your name and point there until we can get the repo moved under `newrelic/`
- Fork and vendor the action under `.github` as you mentioned

Comment (hannahramadan) [reply .github/versions.yml:11]: Looks like Jan 6, 2025 is end of support. I'll remove this and add these comments to the ticket.

Comment (hannahramadan) [reply .github/versions.yml]: Linked this convo to the [kanga333 ticket](https://github.com/orgs/newrelic/projects/84/views/1?pane=issue&itemId=71162310).

**Review by fallwith — APPROVED** [review]

**Pull request merged**
